### DISCUSSION TITLE
2 주차 사전 과제, 정지윤

### DISCUSSION THREAD

**Discussion by wavvvvy** [MrMirror21/2024-KHU-CSE-TrackStudy-Web#25] [General]
### 1번
일상에서 접할 수 있는 **상태**에는 배터리가 있다.

**배터리 부족** ➡️ **충전 중** 
: 배터리 부족 아이콘 또는 상태에서 충전 중 아이콘이 띄워지며 충전 상태로 변함
: 전원을 공급받아 배터리가 충전되게 됨

**충전 중** ➡️ **충전 완료** 
: 배터리가 100%가 됐을 때 충전 중에서 충전 완료 상태로 변함
: 전원을 뽑아도 기기를 사용할 수 있음

**충전 완료** ➡️ **배터리 부족**
: 기기를 사용하며 배터리가 소모되면 충전 완료에서 배터리 부족 상태로 변함
: 충전이 필요하여 전원 연결 전까지 배터리 부족 아이콘이 뜨게 됨. 기기의 전원이 꺼질 수 있음

### 2번
**유튜브 메인 화면**에서의 사용자 입력에 따른 반응 요소

1️⃣ 검색어 자동 완성 기능
: 사용자가 텍스트를 **입력**하기 시작하면 사용자의 입력에 따라 실시간으로 추천 검색어와 인기 검색어 목록을 보여줌
2️⃣ 알림 및 메시지 확인 기능
: 상단의 알림 아이콘을 **클릭**하면 사용자의 활동에 대한 실시간 알림을 확인할 수 있음
3️⃣ 동영상 썸네일로 영상 미리보기
: 사용자가 썸네일 위로 마우스를 올리면, **onmouseover** 이벤트에 의해 미리보기 영상이 로드됨

### 3번
```console.log(x.count)```의 결과값은 **1**
```let y = x``` 로 y가 x의 참조인 ```count:0``` 을 동일하게 참조하게 되어 ```y.count = 1``` 로 속성 값을 변경했을 때, x도 영향을 받아 ```console.log(x.count)``` 는 **1**이 출력된다. 

### 4번
때에 따라서는 불변성이 중요하지 않다고 생각한다. 불변 데이터를 사용하면 매번 새로운 객체를 생성하게 되어 메모리의 사용량이 증가하고 코드의 복잡성이 증가하게 되기 때문이다. 실시간 데이터 처리를 하는 경우가 이에 해당된다. 

### 5번
```
시작
끝
1초후
3초후
``` 
출력을 예상 하였고 예상과 결과는 동일

`console.log` 는 즉시 실행되는 동기 코드이기에 `시작`과 `끝`은 바로 출력이 되는 반면, 
`setTimeout` 함수는 비동기 함수로, 지정된 시간이 지난 후 콜백 함수를 실행하기에 각 1초와 3초 후 `1초` 와 `3초` 라는 결과가 출력되게 된다.

**Comment by MrMirror21** [comment]
작성해주신 사전과제 잘 봤습니다. 전체적으로 답변 훌륭하게 잘 답변해주셨고, 풀이형식 문제도 잘 풀어주신 것 같습니다. 불변성을 준수할 때의 단점 역시 잘 지적해주셨네요. 사전과제 작성이 지난 3주차 스터디 내용을 이해하는데 도움이 되셨으면 좋겠습니다. 열심히 참여해주셔서 감사합니다 🙌
